File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/ObjectShape.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.UUID;

/**
 * ObjectShape
 *
 * <p>Template: {@code ObjectShape Low 98 NotTrusted Zerocoded}
 * (recovered/reference/message_template.msg).
 */
public class ObjectShape extends SLMessage {
    public AgentData AgentData_Field;
    public ArrayList<ObjectData> ObjectData_Fields = new ArrayList<>();

    /** Block AgentData, Single. */
    public static class AgentData {
        public UUID AgentID; // LLUUID
        public UUID SessionID; // LLUUID
    }

    /** Block ObjectData, Variable. */
    public static class ObjectData {
        public int ObjectLocalID; // U32
        public int PathBegin; // U16 - 0 to 1, quanta = 0.01
        public int PathCurve; // U8
        public int PathEnd; // U16 - 0 to 1, quanta = 0.01
        public int PathRadiusOffset; // S8 - -1 to 1, quanta = 0.01
        public int PathRevolutions; // U8 - 0 to 3, quanta = 0.015
        public int PathScaleX; // U8 - 0 to 1, quanta = 0.01
        public int PathScaleY; // U8 - 0 to 1, quanta = 0.01
        public int PathShearX; // U8 - -.5 to .5, quanta = 0.01
        public int PathShearY; // U8 - -.5 to .5, quanta = 0.01
        public int PathSkew; // S8 - -1 to 1, quanta = 0.01
        public int PathTaperX; // S8 - -1 to 1, quanta = 0.01
        public int PathTaperY; // S8 - -1 to 1, quanta = 0.01
        public int PathTwist; // S8 - -1 to 1, quanta = 0.01
        public int PathTwistBegin; // S8 - -1 to 1, quanta = 0.01
        public int ProfileBegin; // U16 - 0 to 1, quanta = 0.01
        public int ProfileCurve; // U8
        public int ProfileEnd; // U16 - 0 to 1, quanta = 0.01
        public int ProfileHollow; // U16 - 0 to 1, quanta = 0.01
    }

    public ObjectShape() {
        this.zeroCoded = true;
        this.AgentData_Field = new AgentData();
    }

    @Override
    public int CalcPayloadSize() {
        return (this.ObjectData_Fields.size() * 27) + 37;
    }

    @Override
    public void Handle(SLMessageHandler messageHandler) {
        messageHandler.HandleObjectShape(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 98 (ObjectShape).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x00);
        byteBuffer.put((byte) 0x62);
        packUUID(byteBuffer, this.AgentData_Field.AgentID);
        packUUID(byteBuffer, this.AgentData_Field.SessionID);
        byteBuffer.put((byte) this.ObjectData_Fields.size());
        for (ObjectData objectData : this.ObjectData_Fields) {
            packInt(byteBuffer, objectData.ObjectLocalID);
            packByte(byteBuffer, (byte) objectData.PathCurve);
            packByte(byteBuffer, (byte) objectData.ProfileCurve);
            packShort(byteBuffer, (short) objectData.PathBegin);
            packShort(byteBuffer, (short) objectData.PathEnd);
            packByte(byteBuffer, (byte) objectData.PathScaleX);
            packByte(byteBuffer, (byte) objectData.PathScaleY);
            packByte(byteBuffer, (byte) objectData.PathShearX);
            packByte(byteBuffer, (byte) objectData.PathShearY);
            packByte(byteBuffer, (byte) objectData.PathTwist);
            packByte(byteBuffer, (byte) objectData.PathTwistBegin);
            packByte(byteBuffer, (byte) objectData.PathRadiusOffset);
            packByte(byteBuffer, (byte) objectData.PathTaperX);
            packByte(byteBuffer, (byte) objectData.PathTaperY);
            packByte(byteBuffer, (byte) objectData.PathRevolutions);
            packByte(byteBuffer, (byte) objectData.PathSkew);
            packShort(byteBuffer, (short) objectData.ProfileBegin);
            packShort(byteBuffer, (short) objectData.ProfileEnd);
            packShort(byteBuffer, (short) objectData.ProfileHollow);
        }
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.AgentData_Field.AgentID = unpackUUID(byteBuffer);
        this.AgentData_Field.SessionID = unpackUUID(byteBuffer);
        int i = byteBuffer.get() & 0xFF;
        for (int j = 0; j < i; j++) {
            ObjectData objectData = new ObjectData();
            objectData.ObjectLocalID = unpackInt(byteBuffer);
            objectData.PathCurve = unpackByte(byteBuffer) & 0xFF;
            objectData.ProfileCurve = unpackByte(byteBuffer) & 0xFF;
            objectData.PathBegin = unpackShort(byteBuffer) & 65535;
            objectData.PathEnd = unpackShort(byteBuffer) & 65535;
            objectData.PathScaleX = unpackByte(byteBuffer) & 0xFF;
            objectData.PathScaleY = unpackByte(byteBuffer) & 0xFF;
            objectData.PathShearX = unpackByte(byteBuffer) & 0xFF;
            objectData.PathShearY = unpackByte(byteBuffer) & 0xFF;
            objectData.PathTwist = unpackByte(byteBuffer);
            objectData.PathTwistBegin = unpackByte(byteBuffer);
            objectData.PathRadiusOffset = unpackByte(byteBuffer);
            objectData.PathTaperX = unpackByte(byteBuffer);
            objectData.PathTaperY = unpackByte(byteBuffer);
            objectData.PathRevolutions = unpackByte(byteBuffer) & 0xFF;
            objectData.PathSkew = unpackByte(byteBuffer);
            objectData.ProfileBegin = unpackShort(byteBuffer) & 65535;
            objectData.ProfileEnd = unpackShort(byteBuffer) & 65535;
            objectData.ProfileHollow = unpackShort(byteBuffer) & 65535;
            this.ObjectData_Fields.add(objectData);
        }
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/ObjectShape.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer
import java.util.ArrayList
import java.util.UUID

/**
 * ObjectShape
 *
 * <p>Template: {@code ObjectShape Low 98 NotTrusted Zerocoded}
 * (recovered/reference/message_template.msg).
 */
open class ObjectShape : SLMessage() {
    @JvmField var AgentData_Field: AgentData = AgentData()
    @JvmField val ObjectData_Fields = ArrayList<ObjectData>()

    /** Block AgentData, Single. */
    open class AgentData {
        @JvmField var AgentID: UUID? = null
        @JvmField var SessionID: UUID? = null
    }

    /** Block ObjectData, Variable. */
    open class ObjectData {
        @JvmField var ObjectLocalID: Int = 0
        @JvmField var PathBegin: Int = 0
        @JvmField var PathCurve: Int = 0
        @JvmField var PathEnd: Int = 0
        @JvmField var PathRadiusOffset: Int = 0
        @JvmField var PathRevolutions: Int = 0
        @JvmField var PathScaleX: Int = 0
        @JvmField var PathScaleY: Int = 0
        @JvmField var PathShearX: Int = 0
        @JvmField var PathShearY: Int = 0
        @JvmField var PathSkew: Int = 0
        @JvmField var PathTaperX: Int = 0
        @JvmField var PathTaperY: Int = 0
        @JvmField var PathTwist: Int = 0
        @JvmField var PathTwistBegin: Int = 0
        @JvmField var ProfileBegin: Int = 0
        @JvmField var ProfileCurve: Int = 0
        @JvmField var ProfileEnd: Int = 0
        @JvmField var ProfileHollow: Int = 0
    }

    init {
        zeroCoded = true
    }

    override fun CalcPayloadSize(): Int {
        return (ObjectData_Fields.size * 27) + 37
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleObjectShape(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 98 (ObjectShape).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x00).toByte())
        byteBuffer.put((0x62).toByte())
        packUUID(byteBuffer, AgentData_Field.AgentID)
        packUUID(byteBuffer, AgentData_Field.SessionID)
        byteBuffer.put((ObjectData_Fields.size.toByte()))
        for (objectData in ObjectData_Fields) {
            packInt(byteBuffer, objectData.ObjectLocalID)
            packByte(byteBuffer, (objectData.PathCurve).toByte())
            packByte(byteBuffer, (objectData.ProfileCurve).toByte())
            packShort(byteBuffer, objectData.PathBegin.toShort())
            packShort(byteBuffer, objectData.PathEnd.toShort())
            packByte(byteBuffer, (objectData.PathScaleX).toByte())
            packByte(byteBuffer, (objectData.PathScaleY).toByte())
            packByte(byteBuffer, (objectData.PathShearX).toByte())
            packByte(byteBuffer, (objectData.PathShearY).toByte())
            packByte(byteBuffer, (objectData.PathTwist).toByte())
            packByte(byteBuffer, (objectData.PathTwistBegin).toByte())
            packByte(byteBuffer, (objectData.PathRadiusOffset).toByte())
            packByte(byteBuffer, (objectData.PathTaperX).toByte())
            packByte(byteBuffer, (objectData.PathTaperY).toByte())
            packByte(byteBuffer, (objectData.PathRevolutions).toByte())
            packByte(byteBuffer, (objectData.PathSkew).toByte())
            packShort(byteBuffer, objectData.ProfileBegin.toShort())
            packShort(byteBuffer, objectData.ProfileEnd.toShort())
            packShort(byteBuffer, objectData.ProfileHollow.toShort())
        }
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        AgentData_Field.AgentID = unpackUUID(byteBuffer)
        AgentData_Field.SessionID = unpackUUID(byteBuffer)
        val i = (byteBuffer.get().toInt() and 0xFF)
        repeat(i) {
            val objectData = ObjectData()
            objectData.ObjectLocalID = unpackInt(byteBuffer)
            objectData.PathCurve = unpackByte(byteBuffer).toInt() and 0xFF
            objectData.ProfileCurve = unpackByte(byteBuffer).toInt() and 0xFF
            objectData.PathBegin = unpackShort(byteBuffer).toInt() and 65535
            objectData.PathEnd = unpackShort(byteBuffer).toInt() and 65535
            objectData.PathScaleX = unpackByte(byteBuffer).toInt() and 0xFF
            objectData.PathScaleY = unpackByte(byteBuffer).toInt() and 0xFF
            objectData.PathShearX = unpackByte(byteBuffer).toInt() and 0xFF
            objectData.PathShearY = unpackByte(byteBuffer).toInt() and 0xFF
            objectData.PathTwist = unpackByte(byteBuffer).toInt()
            objectData.PathTwistBegin = unpackByte(byteBuffer).toInt()
            objectData.PathRadiusOffset = unpackByte(byteBuffer).toInt()
            objectData.PathTaperX = unpackByte(byteBuffer).toInt()
            objectData.PathTaperY = unpackByte(byteBuffer).toInt()
            objectData.PathRevolutions = unpackByte(byteBuffer).toInt() and 0xFF
            objectData.PathSkew = unpackByte(byteBuffer).toInt()
            objectData.ProfileBegin = unpackShort(byteBuffer).toInt() and 65535
            objectData.ProfileEnd = unpackShort(byteBuffer).toInt() and 65535
            objectData.ProfileHollow = unpackShort(byteBuffer).toInt() and 65535
            ObjectData_Fields.add(objectData)
        }
    }
}
